File: 543. Diameter of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = Integer.MIN_VALUE;
    public int getAns(TreeNode node){
        if(node == null) return 0;
        int left = getAns(node.left);
        int right = getAns(node.right);
        
        ans = Math.max(ans, left + right);
        // System.out.println(left + " " + right + " : " + node.val);
        return 1 + Math.max(left, right);
    }

    public int diameterOfBinaryTree(TreeNode root) {
        getAns(root);
        return ans;    
    }
}

File: 543. Diameter of Binary Tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *leftNode, TreeNode *rightNode)
//         : val(x), left(leftNode), right(rightNode) {}
// };

class Solution {
public:
    int ans = INT_MIN;
 
    int find(TreeNode* root) {
        if (root == nullptr) return 0;
        int left = find(root->left);
        int right = find(root->right);

        ans = max(ans, 1 + left + right);
        return 1 + max(left, right);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        find(root);
        return ans - 1;
    }
};
